File: src/table-compare/mineru-client.ts
import { readFile } from "node:fs/promises";
import path from "node:path";

import type { MinerUOptions } from "./types";

const DEFAULT_OPTIONS: Required<MinerUOptions> = {
  lang: "en",
  backend: "hybrid-auto-engine",
  parseMethod: "auto",
  formulaEnable: true,
  tableEnable: true,
  imageAnalysis: true,
  startPageId: 0,
  endPageId: 99999,
};

export interface MinerUClientConfig {
  baseUrl: string;
  pollIntervalMs?: number;
  resultTimeoutMs?: number;
}

export class MinerUClient {
  private readonly baseUrl: string;
  private readonly pollIntervalMs: number;
  private readonly resultTimeoutMs: number;

  constructor(config: MinerUClientConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, "");
    this.pollIntervalMs = config.pollIntervalMs ?? 2_000;
    this.resultTimeoutMs = config.resultTimeoutMs ?? 30 * 60_000;
  }

  async health(): Promise<unknown> {
    const response = await fetch(`${this.baseUrl}/health`);
    if (!response.ok) {
      throw new Error(`MinerU health failed: ${response.status} ${await response.text()}`);
    }
    return response.json();
  }

  async parseDocument(filePath: string, options: MinerUOptions = {}): Promise<{ taskId: string; result: unknown }> {
    const merged = { ...DEFAULT_OPTIONS, ...options };
    const fileName = path.basename(filePath);
    const form = new FormData();
    const bytes = await readFile(filePath);

    form.append("files", new Blob([bytes], { type: guessMimeType(fileName) }), fileName);
    form.append("lang_list", merged.lang);
    form.append("backend", merged.backend);
    form.append("parse_method", merged.parseMethod);
    form.append("formula_enable", String(merged.formulaEnable));
    form.append("table_enable", String(merged.tableEnable));
    form.append("image_analysis", String(merged.imageAnalysis));
    form.append("return_md", "true");
    form.append("return_middle_json", "true");
    form.append("return_model_output", "false");
    form.append("return_content_list", "true");
    form.append("return_images", "false");
    form.append("response_format_zip", "false");
    form.append("start_page_id", String(merged.startPageId));
    form.append("end_page_id", String(merged.endPageId));

    const submit = await fetch(`${this.baseUrl}/tasks`, {
      method: "POST",
      body: form,
    });
    if (!submit.ok) {
      throw new Error(`MinerU task submit failed: ${submit.status} ${await submit.text()}`);
    }

    const submitted = (await submit.json()) as { task_id?: string; taskId?: string };
    const taskId = submitted.task_id ?? submitted.taskId;
    if (!taskId) {
      throw new Error(`MinerU response did not include a task id: ${JSON.stringify(submitted)}`);
    }

    const result = await this.waitForResult(taskId);
    return { taskId, result };
  }

  private async waitForResult(taskId: string): Promise<unknown> {
    const deadline = Date.now() + this.resultTimeoutMs;
    let lastStatus: unknown;

    while (Date.now() < deadline) {
      const statusResponse = await fetch(`${this.baseUrl}/tasks/${taskId}`);
      if (!statusResponse.ok) {
        throw new Error(`MinerU task status failed: ${statusResponse.status} ${await statusResponse.text()}`);
      }
      const status = (await statusResponse.json()) as { status?: string; error?: string };
      lastStatus = status;

      if (status.status === "completed") {
        const resultResponse = await fetch(`${this.baseUrl}/tasks/${taskId}/result`);
        if (!resultResponse.ok) {
          throw new Error(`MinerU task result failed: ${resultResponse.status} ${await resultResponse.text()}`);
        }
        return resultResponse.json();
      }
      if (status.status === "failed") {
        throw new Error(status.error ?? `MinerU task ${taskId} failed`);
      }
      await new Promise((resolve) => setTimeout(resolve, this.pollIntervalMs));
    }

    throw new Error(`Timed out waiting for MinerU task ${taskId}: ${JSON.stringify(lastStatus)}`);
  }
}

function guessMimeType(fileName: string): string {
  const lower = fileName.toLowerCase();
  if (lower.endsWith(".pdf")) {
    return "application/pdf";
  }
  if (lower.endsWith(".doc")) {
    return "application/msword";
  }
  if (lower.endsWith(".docx")) {
    return "application/vnd.openxmlformats-officedocument.wordprocessingml.document";
  }
  if (lower.endsWith(".png")) {
    return "image/png";
  }
  if (lower.endsWith(".jpg") || lower.endsWith(".jpeg")) {
    return "image/jpeg";
  }
  if (lower.endsWith(".webp")) {
    return "image/webp";
  }
  return "application/octet-stream";
}

File: src/pdfparse_agent/mineru_client.py
import asyncio
import mimetypes
from pathlib import Path
from time import monotonic
from typing import Any

import httpx

from .models import ParseOptions


class MinerUClient:
    def __init__(
        self,
        base_url: str,
        timeout_seconds: float,
        poll_interval_seconds: float,
        result_timeout_seconds: float,
    ):
        self.base_url = base_url.rstrip("/")
        self.poll_interval_seconds = poll_interval_seconds
        self.result_timeout_seconds = result_timeout_seconds
        self.client = httpx.AsyncClient(
            base_url=self.base_url,
            timeout=httpx.Timeout(timeout_seconds, read=timeout_seconds),
        )

    async def close(self) -> None:
        await self.client.aclose()

    async def health(self) -> dict[str, Any]:
        response = await self.client.get("/health")
        response.raise_for_status()
        return response.json()

    async def submit_task(self, paths: list[Path], options: ParseOptions) -> dict[str, Any]:
        files = []
        handles = []
        try:
            for path in paths:
                media_type = mimetypes.guess_type(path.name)[0] or "application/pdf"
                handle = path.open("rb")
                handles.append(handle)
                files.append(("files", (path.name, handle, media_type)))

            data = {
                "lang_list": options.lang,
                "backend": options.backend,
                "parse_method": options.parse_method,
                "formula_enable": str(options.formula_enable).lower(),
                "table_enable": str(options.table_enable).lower(),
                "image_analysis": str(options.image_analysis).lower(),
                "return_md": str(options.return_md).lower(),
                "return_middle_json": str(options.return_middle_json).lower(),
                "return_model_output": str(options.return_model_output).lower(),
                "return_content_list": str(options.return_content_list).lower(),
                "return_images": str(options.return_images).lower(),
                "response_format_zip": str(options.response_format_zip).lower(),
                "start_page_id": str(options.start_page_id),
                "end_page_id": str(options.end_page_id),
            }
            response = await self.client.post("/tasks", data=data, files=files)
            response.raise_for_status()
            return response.json()
        finally:
            for handle in handles:
                handle.close()

    async def wait_for_result(self, task_id: str) -> dict[str, Any]:
        deadline = monotonic() + self.result_timeout_seconds
        last_status: dict[str, Any] | None = None
        while monotonic() < deadline:
            status_response = await self.client.get(f"/tasks/{task_id}")
            status_response.raise_for_status()
            last_status = status_response.json()
            status = last_status.get("status")
            if status == "completed":
                result_response = await self.client.get(f"/tasks/{task_id}/result")
                result_response.raise_for_status()
                return result_response.json()
            if status == "failed":
                raise RuntimeError(last_status.get("error") or "MinerU task failed")
            await asyncio.sleep(self.poll_interval_seconds)
        raise TimeoutError(f"Timed out waiting for MinerU task {task_id}: {last_status}")

